feat: persistent lists

ConnorBaker opened the pull request:
<!--

IMPORTANT

Nix is a non-trivial project, so for your contribution to be successful,
it really is important to follow the contributing guidelines:

https://github.com/NixOS/nix/blob/master/CONTRIBUTING.md

Even if you've contributed to open source before, take a moment to read it,
so you understand the process and the expectations.

- what information to include in commit messages
- proper attribution
- volunteering contributions effectively
- how to get help and our review process.

-->

> [!IMPORTANT]
> This PR is a work in progress. I may force-push.

Experimental work on switching to persistent lists via [immer](https://github.com/arximboldi/immer) (thanks for the library suggestion @Mic92!).

Current blockers:

- A printing test fails (see TODOs in `src/libexpr/print-ambiguous.cc`), but only when run through `nix-instantiate --eval` -- `nix eval` returns the correct result. That's odd and makes me think the failing test is symptomatic of a larger problem I introduced.
- I'm unable to evaluate my own system closure because an uninitialized value sneaks in. To help debug this, I've a `printError` to `Value::type` (see `src/libexpr/value.hh`), which is what was throwing the error. I've also added an assert to `EvalState::forceValue` (see `src/libexpr/eval-inline.hh`) to ensure an error is thrown if an uninitialized value is forced.
- I removed tests for the nix_api because I don't know how (or with what) to replace them.

# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

As far as I can tell, lists in Nix are structures with raw C arrays and a size. Most list operations involve allocating an array large enough to hold the result and then `memcpy`-ing components over.

This PR looks at using persistent lists via [immer](https://github.com/arximboldi/immer) -- my hope is that by leveraging techniques from function programming, we can reduce the amount of allocation and copying we do and improve performance of the interpreter.

# Context
<!-- Provide context. Reference open issues if available. -->

TODO

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

roberth commented on src/libexpr/value.hh:
```diff
@@ -13,6 +13,21 @@
+// declare a memory policy for using a tracing garbage collector
+using gc_policy = immer::memory_policy<immer::heap_policy<immer::gc_heap>,
+                                       immer::no_refcount_policy,
+                                       immer::default_lock_policy,
```
I wonder what the default lock policy does.
I suppose clever locking might save allocations in a few cases, but I'm inclined to think that just allocating all the time is more efficient, because we tend to keep the "old version" around because purely functional operations are non-destructive and our scopes/closures (`Env`) tend to stick around fairly often in practice, and if they don't, we don't really know locally.

arximboldi replied:
The locking is not used for allocation actually, but it's used by `immer::atom` so no need to worry about it :)

roberth commented:
> * I'm unable to evaluate my own system closure because an uninitialized value sneaks in.

Memory management bugs are tricky to debug, because their effects tend to be discovered only later on, and in rather uninformative ways.

The warnings in https://sinusoid.es/immer/memory.html#garbage-collected-heap hold true, but I wouldn't be surprised if another property is involved in our crash here.

Does `immer` require internal pointers to be enabled? If not documented, perhaps just change our call to `GC_set_all_interior_pointers`.

Does this work if GC is disabled? If so, we might be dealing with a missing GC root. If it also fails with GC disabled, we might be looking at some sort of memory corruption, such as the lack of initialization you suggested.

roberth reviewed (commented):
Things that jumped out browsing through the code (not a review)

Mic92 commented:
Thanks for doing this experiment. Maybe to see if this goes in the right direction you could try to do some micro benchmark before having to fix all builtins etc? Than we could already get a performance preview.

Mic92 commented:
also cc @arximboldi

ConnorBaker commented:
As of 2472581351eebbb8db1cd676ae0bcb91ac40bc28 I can eval my NixOS config!

I tried re-adding uses of `transient`, but ran into the same errors in `Value::type` where `internalType` has some garbage value. However, running with `GC_DONT_GC=1` allowed the calls to `transient` to work! (Unfortunately, setting `GC_set_all_interior_pointers(1)` did not.) @roberth do have anything you'd recommend reading in terms of GC roots? Sadly, I'm largely unaware of how modern GC works.

ConnorBaker commented:
Rebased to fix merge conflicts.

roberth commented:
> do have anything you'd recommend reading

No reading list unfortunately; except perhaps for the bdwgc docs, which are mostly in its headers. It's reference documentation, not really learning material.

The symptoms you describe suggest that something is not getting marked and then collected by accident, or in other words, some heap objects aren't reachable from the garbage collection roots such as the (main) stack or C++ heap objects allocated with `traceable_allocator`.
We often create the latter with `allocRootValue`.
For STL collections, we generally just pass the allocator along in the template arguments.

Mic92 commented:
Adding `NIX_SHOW_STATS_PATH=/tmp/trace NIX_SHOW_STATS=1` to the benchmark would also allow to get allocation statistics. So it would be interesting if we don't improve CPU time if we at least save memory.

ConnorBaker commented:
@Mic92 I added stats to the numbers section.

@roberth I know that with Haskell, one can use allocation stats provided by GHC's RTS as a weak proxy for performance -- any idea if that's the case with Nix as well? My understanding was that a non-trivial amount of time is spent in the garbage collector.

ConnorBaker commented:
@arximboldi I'm having trouble with transient data structures -- do you have any insight?

In `eval.cc` I'm writing this:

```cpp
void ExprList::eval(EvalState & state, Env & env, Value & v)
{
    auto list = state.allocList();
    for (auto & i : elems)
        *list = list->push_back(i->maybeThunk(state, env));
    v.mkList(list);
}
```

I've tried to rewrite it to use `transient`:

```cpp
void ExprList::eval(EvalState & state, Env & env, Value & v)
{
    auto list = state.allocList();
    auto transient = list->transient();
    for (auto & i : elems) {
        transient.push_back(i->maybeThunk(state, env));
    }
    *list = transient.persistent();
    v.mkList(list);
}
```

Unfortunately, when I do that I get an exception when calling `transient.push_back`!

```console
nix: /nix/store/gkghbi5d6849mwsbcdhnqljz2xnjvnis-immer-0.8.1-unstable-2024-09-18/include/immer/transience/gc_transience_policy.hpp:95: ownee &immer::gc_transience_policy::apply<immer::heap_policy<immer::gc_heap>>::type::ownee::operator=(edit) [HeapPolicy = immer::heap_policy<immer::gc_heap>]: Assertion `e != noone' failed.
```

I believe *something* is being garbage collected when it shouldn't be, and verified this by checking that running with `GC_DONT_GC=1` succeeds.

I know that in `value.hh`, we have typedefs for STL containers to make sure that we're using containers which allocate with `traceable_allocator`:

```cpp
typedef std::vector<Value *, traceable_allocator<Value *>> ValueVector;
typedef std::unordered_map<Symbol, Value *, std::hash<Symbol>, std::equal_to<Symbol>, traceable_allocator<std::pair<const Symbol, Value *>>> ValueMap;
typedef std::map<Symbol, ValueVector, std::less<Symbol>, traceable_allocator<std::pair<const Symbol, ValueVector>>> ValueVectorMap;
```

Do I need to do something similar for `flex_vector_transient`?

Do you have any ideas on what I can do to troubleshoot or debug this? From the error, I'm not entirely sure if it's the `Value` which is being collected, or the nodes belonging to `transient`.

roberth commented:
> one can use allocation stats provided by GHC's RTS as a weak proxy for performance

To my understanding, this is a consequence of widespread use of immutable data structures.
GHC will optimize away some allocations, which Nix does not, so by that token Nix heap allocations may be a better proxy for performance than GHC's allocations.

> My understanding was that a non-trivial amount of time is spent in the garbage collector.

Recent versions of Nix will tell you how much time is spent in GC. Evaluating my NixOS config it spends 10% of time in GC, but an individual package might only spend 1.5% in GC. (anecdata)

Mic92 commented:
So currently we are actually increasing memory usage? I wonder what happens if we remove the optimization from nix for 1,2 element lists from the current implementation. Maybe the comparison is than fairer?

roberth commented:
Arrays are pretty compact, fwiw.

Maybe this optimization would pay off if we had constant folding? IIRC we haven't previously done constant folding because it didn't pay off, but perhaps that was because of all the copying.
Don't know, but might be worth a shot.

Another possibility is that the data structure can't be used efficiently because it has to assume that multiple references to it exist, even though those references might be all from the garbage.
And that's in the best case; https://github.com/NixOS/nix/issues/8285 causes reachable references to linger.

arximboldi commented:
Hi @ConnorBaker !

Thanks a lot for this contribution. I am a fan of Nix and use it in all my systems, so this gets me very excited! :tada: :smile: 

Taking a look at the code, the two things I can see could improve:

1. Use of transients. I see that you are trying that already.

2. I see a few places where access into the data structure happens by index. It seems maybe even we may be even iterating by index in some cases depending on how some of the interfaces are used. The preferred ways of iterating over Immer data-structures are:
  a) `immer::for_each`, definitelly the fastest way!
  b) Where that's not possible, use iterators, still significantly faster than using random access...

As to why are those crashes happening... you mention that the values are being inappropriately GC'ed. You are using the `immer::gc_policy` that assumes that you're using Bohem's GC (conservative tracing GC). I suspect you probably need to provide a custom heap policy that properly interacts with whatever GC Nix is using? Are there any docs/resources on Nix's GC I could maybe take a look at?

arximboldi commented:
Ok, just took a look and it seems we are indeed using Boehm's GC here. And the the implementation of `traceable_allocator` and it just forwards into `GC_MALLOC_UNCOLLECTABLE`... That somewhat does not seem right for this particular use-case, unless we combine it with reference counting. 

I see instances of something that resembles refcounting in the code though (`nix_gc_incref, etc.). That could be integrated via the memory policy.

I think it'd be nice for someone with a bit more knowledge about how the lifetimes of C++ values is managed in Nix and how it all interacts with the GC. It's unclear to me: which allocations are should be made with Boehm's GC `GC_MALLOC`?

Also, what's `nix_gc_incref/decref`? From looking at the implementation, I suspect it's for FFI, in orderto keep objects alive outside of what Boehm's GC can see?

arximboldi commented:
Looking at it again... I wouldn't fully discard that there may be a bug in Immer's use of `gc_transient` policy, which is less battle tested than the refcounting version. I will try find some time this week to investigate...

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-it-possible-for-us-to-remove-builtins-functionargs/51960/4

ConnorBaker commented:
I’ve stalled a bit here in terms of will to make progress.

I’ve been thinking about implementations for the data structures we use for lists and attribute sets, and though this is naive, I would think the best implementation would make the most common operations the cheapest to perform.

Seeing as how lists are implemented as arrays of pointers to values instead of values themselves, I don’t know that the overhead involved in persistent data structures is worth what we can get back in sharing.

Having lazy evaluation, I don’t know that it would make sense to have lists contain actual values rather than pointers to them — for most of our usage of the language, would cache locality matter?

Attribute sets are vectors of Attr, but I remember that at some point they were two arrays of values — one for names, and one for values. Given that outside attribute set merging common patterns of use involve iterating over the names or values of attribute sets, would it make sense for the implementation to make such operations cheap? As I understand it, getting the names or values involved iterating over the vector of Attr and storing the desired components in a new collection.

Not related to lists or attribute sets — strings! I understand the SymbolTable is doing something like string interning? Is there a better data structure we could use rather than a vector? @tomberek in #9034 you mentioned using the symbol table for all strings; are you aware of any existing library we could use for more efficient string implementations or operations, or should we look into implementing our own?

arximboldi commented:
Thank you @ConnorBaker for investigating this issue! I'd be happy to accept the patch you mentioned upstream in Immer. I'm taking a look to see if I notice some ways in which performance could be improved. Thanks!

arximboldi commented on src/libexpr/value.hh:
```diff
@@ -244,11 +230,7 @@ public:
-            Value * const * elems;
-        } bigList;
-        Value * smallList[2];
+        ValueList * list;
```
Why `ValueList*` instead of plain `ValueList`? Is it because it's too big (4 vs 2 pointers). I wonder how much of a difference this would make in the benchmarks...

arximboldi replied:
Yes, I meant the `list` itself being `ValueList` instead of `ValueList*`, not the `Value*`.

xokdvium replied:
Yeah I noted below about bumping the size of Value. Sorry for the confusion :)

arximboldi commented on src/nix-env/user-env.cc:
```diff
@@ -71,9 +71,11 @@ bool createUserEnv(EvalState & state, PackageInfos & elems,
+        for (const auto & j : outputs) {
+            auto v = state.allocValue();
+            v->mkString(j.first);
+            *outputsList = outputsList->push_back(v);
```
Can use transients for better performance...

arximboldi commented:
Sorry for the spam in the comments inline... I realized at some point that I wasn't being useful. I was wondering: are transients still not working? Looking at the current implementation, I would expect performance to improve by a lot simply by using transients. Note that with `gc_policy` doing `x = std::move(x).push_back(...)` is not enough, you have to actually use `x.transient()` to create the transient object and do the iterative `push_back()` on that.

The other big question is whether it could be accepted to increase the size of `Value` so that we can get rid of the extra indirection.

Even with the two issues above, I would expect using `immer` vs the previous implementation to still use more memory. Performance could improve particularly in code that uses long lists.

xokdvium commented:
One thing to consider is that lists are much less often used that attribute sets. That's why I focused my work on attribute sets in e.g.:

> size of Value so that we can get rid of the extra indirection.

That would balloon memory usage. When this PR was written it was 24 bytes, but now it's 16 and adding anything would bring it up back to 24 due to padding. That would also get make Boehm not use the thread local small object allocator that allocate a single granules of 16 bytes (that's my best understanding of boehm's workings).

We now do aggressive pointer tagging to make sure that Value is as small as it can be. Sadly due to the semantic difference of pointer identity the layout can't be changed much (like shrinking to 8 bytes and moving more things on the heap) without also changing evaluation results. For context:

> https://github.com/NixOS/nix/pull/13407

ConnorBaker commented:
I force-pushed and updated https://github.com/ConnorBaker/nix/tree/vibe-coding/use-immer (again, no PR for that branch). It should be using transients where possible, but now only uses Immer for large attribute sets (otherwise the overhead of CHAMP outweighs any possible benefits sharing would provide). It's still slower and uses more memory than the current implementation, but... perhaps that's unavoidable? Is such the nature of tailored solutions?

For what it's worth, I've been comparing results using repeated NixOS system closure evaluation -- `NIX_SHOW_STATS=1 ./result/bin/nix eval --json --file /home/connorbaker/nixpkgs/nixos/release.nix closures`. Since each closure imports and instantiates a copy of Nixpkgs *and* uses the module system, it's a good way to put the evaluator through its paces. (Also good to test memory usage since Nix isn't able to free the instances of Nixpkgs it creates for each system closures.)

ConnorBaker commented:
Thank for pointing out issues with the way I'm using Immer! My experience with C++ is minimal and Immer even less -- having guidance like that is really helpful and I appreciate the feedback :)

I added some extra instrumentation on a separate branch where I was looking into whether it's beneficial to add more special cases for certain builtins (e.g., attribute set of union and empty attribute set is the other attribute set, no need to call lexicographic sorting or other functions). The results of doing some system evaluations (https://gist.github.com/ConnorBaker/610a704e2e4b7f55ad138eeba0e7a688) show the distribution of argument sizes to a number of builtins. It looks the like majority of arguments given to the builtins I added instrumentation to are small (less than 10 elements). (I am realizing now I don't think I added instrumentation for attribute set updates, which is a major oversight on my part.)

Given the distribution for that (perhaps not representative) evaluation, would it still make sense to use Immer?

I know that the evaluator has a small-list optimization where it stores lists of less than three elements directly in the Value rather than allocating a separate list. (I don't think attribute sets have something similar.) Would it make sense to keep that optimization or switch over fully to Immer for lists?

Additionally, would you recommend we store values directly in the data structure instead of storing pointers to values? I'm unfamiliar with how Immer implements structural sharing, and (I think) we use pointers to avoid forcing the same thunk multiple times, so if we were to switch to returning a pointer to a `Value` stored in an Immer data structure, would it be stable? 

On the subject of tuning values like the B parameter, do you have any references or suggestions of values to try?

Again, thank you for the help and thank you for making Immer!

ConnorBaker commented:
I just force-pushed to that same branch (https://github.com/ConnorBaker/nix/tree/vibe-coding/use-immer) and split it into three commits -- bringing in Immer as a Nix dependency and the work to migrate and a fixup afterward.

I apologize in advance -- still have no idea what I'm doing.

Here's a benchmark run using rev `644be074e10dfa2463cf20788d8428b4cc652d01` (immediately prior to my changes):

```console
$ NIX_SHOW_STATS=1 ./nix-old/bin/nix eval --json --file ../nixpkgs/nixos/release.nix closures
{
  "ec2": {
    "aarch64-linux": "/nix/store/sqmnwkf212b3470kgfh523dqi99km22k-nixos-system-unnamed-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/4b149glh0s093zcilyjqw8gjjdcxq8nr-nixos-system-unnamed-26.05pre708350.gfedcba"
  },
  "gnome": {
    "aarch64-linux": "/nix/store/5bd57ijd67ia4bm216rp1fb2hc3p7zc7-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/2cdbgix9yqqpxgqbmrl95vc6hm4abc4p-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "kde": {
    "aarch64-linux": "/nix/store/ri98m314wz7sjj01fiacv5dq280c3j0y-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/nxil2bscxd3kmcpidlnadhg19519s4rk-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "lapp": {
    "aarch64-linux": "/nix/store/5872qqnnml21jq7hyqz47hlbiy5xwr9x-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/fi1v0i0fc1yx92vk66d8ba30jdw54x8v-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "pantheon": {
    "aarch64-linux": "/nix/store/mdrikfy05sb2fw475h4dfl4088y7slfm-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/lkcq5ry7h4xmkpsihgxm3iwnb78y1cvs-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "smallContainer": {
    "aarch64-linux": "/nix/store/sfng16666wa9l9xcdklhfqgv7djiynp0-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/pzcvd5yn6l1cc39n4y4mim75hnd8jczi-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "tinyContainer": {
    "aarch64-linux": "/nix/store/byh5fqqjgk4z6j8m88by6jxab0xgf1pq-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/3l86p95rmw2bm04cyfcabb9fd6jy72ja-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "xfce": {
    "aarch64-linux": "/nix/store/jj96nsdks3v1h9kj6sg9vqkzj02a77jj-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/1qjhbjyhhhrsn018hh7690fs20pdabpy-nixos-system-nixos-26.05pre708350.gfedcba"
  }
}
{
  "cpuTime": 42.663089752197266,
  "envs": {
    "bytes": 1980546704,
    "elements": 144274177,
    "number": 103294161
  },
  "gc": {
    "cycles": 15,
    "heapSize": 2416181248,
    "totalBytes": 9815425120
  },
  "list": {
    "bytes": 252692528,
    "concats": 3793501,
    "elements": 31586566
  },
  "nrAvoided": 106344563,
  "nrExprs": 1637954,
  "nrFunctionCalls": 89801281,
  "nrLookups": 56478286,
  "nrOpUpdateValuesCopied": 95019450,
  "nrOpUpdates": 11529567,
  "nrPrimOpCalls": 41227302,
  "nrThunks": 126534641,
  "sets": {
    "bytes": 3710039136,
    "elements": 190687377,
    "number": 27460046
  },
  "sizes": {
    "Attr": 16,
    "Bindings": 24,
    "Env": 8,
    "Value": 16
  },
  "symbols": {
    "bytes": 1244036,
    "number": 105439
  },
  "time": {
    "cpu": 42.663089752197266,
    "gc": 1.483,
    "gcFraction": 0.03476072662842291
  },
  "values": {
    "bytes": 2852997440,
    "number": 178312340
  }
}
```

and here's with those changes:

```console
$ NIX_SHOW_STATS=1 ./nix-new/bin/nix eval --json --file ../nixpkgs/nixos/release.nix closures
{
  "ec2": {
    "aarch64-linux": "/nix/store/sqmnwkf212b3470kgfh523dqi99km22k-nixos-system-unnamed-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/4b149glh0s093zcilyjqw8gjjdcxq8nr-nixos-system-unnamed-26.05pre708350.gfedcba"
  },
  "gnome": {
    "aarch64-linux": "/nix/store/5bd57ijd67ia4bm216rp1fb2hc3p7zc7-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/2cdbgix9yqqpxgqbmrl95vc6hm4abc4p-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "kde": {
    "aarch64-linux": "/nix/store/ri98m314wz7sjj01fiacv5dq280c3j0y-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/nxil2bscxd3kmcpidlnadhg19519s4rk-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "lapp": {
    "aarch64-linux": "/nix/store/5872qqnnml21jq7hyqz47hlbiy5xwr9x-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/fi1v0i0fc1yx92vk66d8ba30jdw54x8v-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "pantheon": {
    "aarch64-linux": "/nix/store/mdrikfy05sb2fw475h4dfl4088y7slfm-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/lkcq5ry7h4xmkpsihgxm3iwnb78y1cvs-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "smallContainer": {
    "aarch64-linux": "/nix/store/sfng16666wa9l9xcdklhfqgv7djiynp0-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/pzcvd5yn6l1cc39n4y4mim75hnd8jczi-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "tinyContainer": {
    "aarch64-linux": "/nix/store/byh5fqqjgk4z6j8m88by6jxab0xgf1pq-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/3l86p95rmw2bm04cyfcabb9fd6jy72ja-nixos-system-nixos-26.05pre708350.gfedcba"
  },
  "xfce": {
    "aarch64-linux": "/nix/store/jj96nsdks3v1h9kj6sg9vqkzj02a77jj-nixos-system-nixos-26.05pre708350.gfedcba",
    "x86_64-linux": "/nix/store/1qjhbjyhhhrsn018hh7690fs20pdabpy-nixos-system-nixos-26.05pre708350.gfedcba"
  }
}
{
  "cpuTime": 107.35895538330078,
  "envs": {
    "bytes": 1980573520,
    "elements": 144275893,
    "number": 103295797
  },
  "gc": {
    "cycles": 34,
    "heapSize": 5687738368,
    "totalBytes": 35781190304
  },
  "list": {
    "bytes": 252692784,
    "concats": 3793565,
    "elements": 31586598
  },
  "nrAvoided": 106344691,
  "nrExprs": 1637954,
  "nrFunctionCalls": 89802853,
  "nrLookups": 56478927,
  "nrOpUpdateValuesCopied": 217671148,
  "nrOpUpdates": 11529599,
  "nrPrimOpCalls": 41227302,
  "nrThunks": 126534945,
  "sets": {
    "bytes": 5584368864,
    "elements": 303983022,
    "number": 30026688
  },
  "sizes": {
    "Attr": 16,
    "AttrValue": 16,
    "Bindings": 24,
    "Env": 8,
    "Value": 16
  },
  "symbols": {
    "bytes": 1244036,
    "number": 105439
  },
  "time": {
    "cpu": 107.35895538330078,
    "gc": 5.437,
    "gcFraction": 0.05064319022654818
  },
  "values": {
    "bytes": 2853002320,
    "number": 178312645
  }
}
```
